lib/dataset/arithmetic.cpp

```diff
@@ -86,39 +86,41 @@ auto apply_with_broadcast(const Op &op, const A &a, const B &b) {
   Dataset res;
   for (const auto &item : b)
     if (const auto it = a.find(item.name()); it != a.end())
-      res.setData(item.name(), op(*it, item));
+      res.setDataInit(item.name(), op(*it, item));
+  if (!res.is_valid())
+    return Dataset({}, {}); // Make sure to always return a valid dataset.
   return res;
 }
 
 template <class Op, class A>
 auto apply_with_broadcast(const Op &op, const A &a, const DataArray &b) {
   Dataset res;
   for (const auto &item : a)
-    res.setData(item.name(), op(item, b));
-  return res;
+    res.setDataInit(item.name(), op(item, b));
+  return res.is_valid() ? res : Dataset({}, {});
 }
 
 template <class Op, class B>
 auto apply_with_broadcast(const Op &op, const DataArray &a, const B &b) {
   Dataset res;
   for (const auto &item : b)
-    res.setData(item.name(), op(a, item));
+    res.setDataInit(item.name(), op(a, item));
   return res;
 }
```

Comment:
There is a discrepancy here. The latter can return an invalid dataset if the input has no items?

In general, is it the correct approach to return `Dataset({}, {})`? I was first wondering if we should just keep the coords, but that is of course not what the old implementation did. Would it make sense to create a dataset with the same dims? Conceptually:

```c++
Dataset res(merge(a.dims(), b.dims())); // not necessarily this syntax, defines dims
for (const auto &item: a)
  res.setData(item.name(), ...);  // no need for `setDataInit`
return res;  // no risk of returning invalid
```

Reply:
Didn't we conclude that we shouldn't have a constructor that takes sizes?

But this approach seems reasonable provided that we can always predict the output shape of `op(a, item)`. Is it guaranteed to produce the merged dims of the inputs?

Reply:
> Didn't we conclude that we shouldn't have a constructor that takes sizes?

I think we did (at least on the Python side), but I cannot quite remember why. But maybe we need it in C++?

> But this approach seems reasonable provided that we can always predict the output shape of `op(a, item)`. Is it guaranteed to produce the merged dims of the inputs?

I think so, that is what `transform` does?

Reply:
There are cases that are more complicated. E.g.
```.cpp
template <class Func, class... Args>
Dataset apply_to_items(const Dataset &d, Func func, Args &&...args) {
  Dataset result;
  for (const auto &data : d)
    result.setDataInit(data.name(), func(data, std::forward<Args>(args)...));
  return result;
}
```
where we only know the size after applying `func`. So avoiding the current approach would mean handling the first iteration of the loop differently from the rest. `setDataInit` was my attempt at keeping the loops tidy.

Also, in this case, what would be the output shape if the input has no items? Simply returning the input shape could lead to problems downstream if the caller expects a change in dims.

Reply:
So, how can we avoid returning invalid `Dataset`?